[NN] Add EdgeGAT operator

schmidt-ju opened the pull request:
## Description
Hey,

I added the EdgeGAT operator of the Robotics and Automation Letters (RA-L) paper [SCENE: Reasoning about Traffic Scenes using Heterogeneous Graph Neural Networks](https://arxiv.org/abs/2301.03512).
The already available [EGAT](https://docs.dgl.ai/en/0.9.x/generated/dgl.nn.pytorch.conv.EGATConv.html) operator does only include edge feature for the calculation of the attention coefficients.

The operator of this PR, namely EdgeGAT, is more advanced: It also includes the edge features directly for the node update and not only for the calculation of the attention coefficients.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change


Looking forward to hearing from you!


Julian

schmidt-ju commented:
Hey,

I am not sure, why the CI is failing. I just recreated the test cases and they seem to work on my local machine.

Also, I am not able to access the produced logs.
There is probably something that I am missing right now. Can you help me?

Julian

BarclayII commented:
@dgl-bot

BarclayII commented:
@schmidt-ju Hi Julian, currently CI requires authentication from one of the core developers so someone from us has to invoke the CI manually.  And after the CI is invoked and an error other than "Authentication failed" is shown, you will have to click on the `Report path` link and find the logs like below:
![image](https://user-images.githubusercontent.com/2978100/218263204-fc8041fe-fab2-4a91-b5d5-a09b1bff1f76.png)
Looks like there is lint issues.  Could you take a look?

Please feel free to nudge us if you have any problems.  Thanks!

schmidt-ju commented:
Hey,

thanks for the information!
I fixed the lint issues (to the best of my knowledge^^). At least on my local machine the `task_lint.sh` script does not throw errors anymore.

BarclayII commented:
Seems that there's still lint issues.  You could try [`lintrunner -a`](https://pypi.org/project/lintrunner/) to apply the lint fixes automatically.

BarclayII commented:
The files are `python/dgl/nn/pytorch/conv/edgegatconv.py` and `python/dgl/nn/pytorch/conv/__init__.py`.

schmidt-ju commented:
Hey @BarclayII,

I fixed the lint issues in the two mentioned files using lintrunner.
According to the "Checks" tab, there are still some issues in `tests/pytorch/test_nn.py`.

My modifications in this file are limited to two new test methods. The linter, however, shows errors in the whole file and not limited to my small changes.
Is this intentional?
I can reformat the whole file, but I don't think that this is something you want.

Looking forward to hearing from you.

Julian

frozenbugs commented:
@dgl-bot

frozenbugs commented:
@schmidt-ju  Can you fix the conflict, I think the lint issue should be addressed.

schmidt-ju commented:
Hey @frozenbugs,
I resolved the conflicts. Lint issues are now gone, too (because of the changes in the test_nn.py file).

frozenbugs commented:
@dgl-bot

frozenbugs commented on python/dgl/nn/pytorch/conv/edgegatconv.py:
```diff
@@ -0,0 +1,381 @@
+                        graph.number_of_dst_nodes(),
+                    ) + dst_prefix_shape[1:]
+
+            # linearly tranform the edge features
```
super nit: capitalize the L and end with . for all comments.

frozenbugs commented on python/dgl/nn/pytorch/conv/edgegatconv.py:
```diff
@@ -0,0 +1,381 @@
+            graph.apply_edges(fn.u_add_v("el", "er", "e_tmp"))
+
+            # e_tmp combines attention weights of source and destination node
+            # we also add the attention weight of the edge
```
Avoiding usage of "we", "you"? Make it a statement.

frozenbugs commented on python/dgl/nn/pytorch/conv/edgegatconv.py:
```diff
@@ -0,0 +1,381 @@
+from ...functional import edge_softmax
+
+# pylint: enable=W0235
+class EdgeGATConv(nn.Module):
```
Maybe mention SCENE in the name of EdgeGATConv? Since we already have another similar egatconv https://github.com/dmlc/dgl/blob/master/python/dgl/nn/pytorch/conv/egatconv.py

BarclayII commented on python/dgl/nn/pytorch/conv/edgegatconv.py:
```diff
@@ -0,0 +1,381 @@
+    r"""Graph attention layer with edge features from `SCENE
+    <https://arxiv.org/pdf/2301.03512.pdf>`__
+
+    .. math::
```
A blank line is needed under `.. math::` block.

BarclayII commented on python/dgl/nn/pytorch/conv/edgegatconv.py:
```diff
@@ -0,0 +1,381 @@
+    for the transformation of features of the node to update (s=self),
+    neighboring nodes (n=neighbor) and edge features (e=edge).
+    Attention weights are obtained by
+    .. math::
```
Ditto

BarclayII reviewed (approved):
LGTM, we can merge it after the code style comments are addressed.

frozenbugs commented:
Overall LGTM, thanks for using and contributing to DGL. @BarclayII will take another look then approve.

schmidt-ju commented:
Hey @BarclayII,

thanks for your overall positive feedback.

I addressed all of your comments.
The only thing left is @frozenbugs comment:
> Maybe mention SCENE in the name of EdgeGATConv? Since we already have another similar egatconv https://github.com/dmlc/dgl/blob/master/python/dgl/nn/pytorch/conv/egatconv.py

I have talked to my co-author about this and we would like to leave the name as it is.
This makes the name in the paper and in DGL consistent with each other.

Looking forward to hearing from you!


Best

Julian

BarclayII commented:
@dgl-bot

schmidt-ju commented:
Hey @BarclayII,

do you already have a rough estimate of when the merge into the main branch will be carried out?

Julian

frozenbugs commented:
> Hey @BarclayII,
> 
> do you already have a rough estimate of when the merge into the main branch will be carried out?
> 
> Julian

Hi Julian, sorry for the trouble, just merged the PR.
(I am new to github, and I thought you can merge also long as we approve)
